### PR TITLE
WIP - node admission

### PR DESCRIPTION
**Pull request by liggitt**

### CODE REVIEW

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
[test]

File: vendor/k8s.io/kubernetes/pkg/apiserver/resthandler.go [hunk 1]
```diff
@@ -149,9 +160,20 @@ func GetResource(r rest.Getter, e rest.Exporter, scope RequestScope) restful.Rou
 }
 
 // GetResourceWithOptions returns a function that handles retrieving a single resource from a rest.Storage object.
-func GetResourceWithOptions(r rest.GetterWithOptions, scope RequestScope) restful.RouteFunction {
+func GetResourceWithOptions(r rest.GetterWithOptions, scope RequestScope, admit admission.Interface) restful.RouteFunction {
 	return getResourceHandler(scope,
 		func(ctx api.Context, name string, req *restful.Request) (runtime.Object, error) {
+
+			if admit != nil && admit.Handles(admission.Get) {
```
Comment (smarterclayton) [inline vendor/k8s.io/kubernetes/pkg/apiserver/resthandler.go]: Let's get this in 1.6

Comment (smarterclayton) [inline vendor/k8s.io/kubernetes/pkg/apiserver/resthandler.go]: There might even be an open PR on it

File: pkg/authorization/admission/restrictnodes/restrictnodes.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	pods, err := q.pods.Pods(ns).List(everything)
+	if err != nil {
+		return err
+	}
+	for _, pod := range pods {
+		if pod.Spec.NodeName != nodeName {
+			continue
+		}
+		found := false
+		visitSecretNames(pod, func(referencedName string) bool {
```
Comment (smarterclayton) [inline pkg/authorization/admission/restrictnodes/restrictnodes.go]: This is what we could memo-ize if it became necessary - invert this to build a set of everything in the namespace, LRU cache it for 30 seconds (assuming all secret accesses are correlated in time), and if you get a match check one more time.  But may not even be that slow.

File: pkg/authorization/admission/restrictnodes/restrictnodes.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+	if len(a.GetSubresource()) > 0 && a.GetSubresource() != "status" {
+		return nil
+	}
+
+	switch a.GetOperation() {
+	case admission.Create:
+		// only let nodes create mirror pods targeted to themselves
+		object := a.GetObject()
+		if object == nil {
+			// TODO: fail closed?
```
Comment (smarterclayton) [inline pkg/authorization/admission/restrictnodes/restrictnodes.go]: Yes, especially if this is an actual node.

Comment (smarterclayton) [inline pkg/authorization/admission/restrictnodes/restrictnodes.go]: Could have the downside of blowing up the cluster... but i'll take a little security

File: pkg/authorization/admission/restrictnodes/restrictnodes.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+func (q *restrictNodesAdmission) admitNode(nodeName string, a admission.Attributes) error {
+	// Only let nodes modify their own API resource
+	if a.GetName() == nodeName {
+		return nil
+	}
+	return kapierrors.NewForbidden(a.GetResource().GroupResource(), a.GetName(), errors.New("cannot access other nodes"))
+}
+
+func (q *restrictNodesAdmission) admitPod(nodeName string, a admission.Attributes) error {
+	if len(a.GetSubresource()) > 0 && a.GetSubresource() != "status" {
```
Comment (smarterclayton) [inline pkg/authorization/admission/restrictnodes/restrictnodes.go]: I think for any other sub resource you want to fail closed.

**Comment by smarterclayton** [comment]
What are next steps on this (record here so someone can follow along at home)

**Comment by liggitt** [comment]
switching gears to an authorizer we'll union in, and pair with an admission plugin for the pod and node create cases

**Pull request closed without merging**